File: keystore/keystore.go
package main

import (
	"encoding/json"
	"fmt"
	"log"

	"github.com/gorilla/websocket"
)

func get(key string) interface{} {
	c, _, err := websocket.DefaultDialer.Dial("wss://cowyo.com/ws", nil)
	if err != nil {
		log.Fatal("dial:", err)
	}
	defer c.Close()
	done := make(chan struct{})
	var value interface{}
	go func() {
		defer c.Close()
		defer close(done)
		for {
			_, message, err := c.ReadMessage()
			if err != nil {
				// log.Println("read:", err)
				return
			}
			var m struct {
				TextData string
			}
			json.Unmarshal([]byte(message), &m)
			json.Unmarshal([]byte(m.TextData), &value)
			done <- struct{}{}
		}
	}()
	// ask for something
	c.WriteMessage(websocket.TextMessage, []byte(`{"Title":"`+key+`", "UpdateClient":true}`))
	<-done
	return value
}

func set(key string, message interface{}) interface{} {
	c, _, err := websocket.DefaultDialer.Dial("wss://cowyo.com/ws", nil)
	if err != nil {
		log.Fatal("dial:", err)
	}
	defer c.Close()
	done := make(chan struct{})
	var value interface{}
	go func() {
		defer c.Close()
		defer close(done)
		for {
			_, message, err := c.ReadMessage()
			if err != nil {
				// log.Println("read:", err)
				return
			}
			var m struct {
				TextData string
			}
			json.Unmarshal([]byte(message), &m)
			json.Unmarshal([]byte(m.TextData), &value)
			done <- struct{}{}
		}
	}()
	// ask for something
	bJson, _ := json.Marshal(message)
	aa, _ := json.Marshal(map[string]interface{}{
		"Title":        key,
		"UpdateServer": true,
		"TextData":     string(bJson),
	})
	c.WriteMessage(websocket.TextMessage, aa)
	<-done
	return value
}

func main() {
	fmt.Println(get("data"))
	fmt.Println(get("data"))
	m := make(map[string]int)
	m["some string"] = 29
	fmt.Println(set("data2", m))
	fmt.Println(get("data2"))
}

File: keystore/keystore.js
function save(key, value) {
  c = new WebSocket('wss://cowyo.com/ws');
  return new Promise(function (resolve, reject) {
    try {
      c.onopen = function (_) {
        c.send(JSON.stringify({
          TextData: JSON.stringify(value),
          Title: `${key}`,
          UpdateServer: true,
          UpdateClient: false,
        }));
        return resolve(true);
      }
    } catch(e) {
      return reject(e);
    }
  });
}

// save('hello2', 'world');




function get(key) {
 c = new WebSocket('wss://cowyo.com/ws');
 return new Promise(function (resolve, reject) {
   try {
     c.onmessage = function(evt) {
       return resolve(JSON.parse(JSON.parse(evt.data).TextData));
     }
     c.onopen = function (_) {
       c.send(JSON.stringify({
         Title: `${key}`,
         UpdateClient: true,
       }));
     };
   } catch(e) {
     return reject(e);
   }
 });
}
